model/modelprocessor/servicecounter_test.go

```diff
@@ -0,0 +1,239 @@
+		{Name: "apm-agent-go", Version: "1.3"},
+		{Name: "apm-agent-node", Version: "1.2"},
+		{Name: "apm-agent-node", Version: "1.2"},
+		{Name: "apm-agent-java", Version: "1.1"},
+	}
+
+	batch := model.Batch{}
+	for _, service := range services {
+		batch = append(batch, model.APMEvent{Service: service})
+	}
+	err := counter.ProcessBatch(context.Background(), &batch)
+	require.NoError(t, err)
+
+	metrics := gatherMetrics(counter)[1:]
+
+	assert.Equal(t, []agent.Metrics{
+		{
+			Labels: agent.StringMap{
+				{Key: "service.name", Value: "apm-agent-go"},
+				{Key: "service.version", Value: "1.3"},
+			},
+			Samples: map[string]agent.Metric{
+				"processed_events": {
+					Value: 3,
+				},
+			},
+		},
+		{
+			Labels: agent.StringMap{
+				{Key: "service.name", Value: "apm-agent-java"},
+				{Key: "service.version", Value: "1.1"},
+			},
+			Samples: map[string]agent.Metric{
+				"processed_events": {
+					Value: 1,
+				},
+			},
+		},
+		{
+			Labels: agent.StringMap{
+				{Key: "service.name", Value: "apm-agent-node"},
+				{Key: "service.version", Value: "1.2"},
+			},
+			Samples: map[string]agent.Metric{
+				"processed_events": {
+					Value: 2,
+				},
+			},
+		},
+	}, metrics)
+
+	services = []model.Service{
+		{Name: "apm-agent-node", Version: "1.2"},
+		{Name: "apm-agent-java", Version: "1.1"},
+		{Name: "apm-agent-java", Version: "1.1"},
+		{Name: "apm-agent-ruby", Version: "1.0"},
+		{Name: "apm-agent-ruby", Version: "1.0"},
+		{Name: "apm-agent-ruby", Version: "1.0"},
+	}
+
+	batch = model.Batch{}
+	for _, service := range services {
+		batch = append(batch, model.APMEvent{Service: service})
+	}
+	err = counter.ProcessBatch(context.Background(), &batch)
+	require.NoError(t, err)
+
+	metrics = gatherMetrics(counter)[1:]
+
+	assert.Equal(t, []agent.Metrics{
+		{
+			Labels: agent.StringMap{
+				{Key: "service.name", Value: "apm-agent-java"},
+				{Key: "service.version", Value: "1.1"},
+			},
+			Samples: map[string]agent.Metric{
+				"processed_events": {
+					Value: 2,
+				},
+			},
+		},
+		{
+			Labels: agent.StringMap{
+				{Key: "service.name", Value: "apm-agent-node"},
+				{Key: "service.version", Value: "1.2"},
+			},
+			Samples: map[string]agent.Metric{
+				"processed_events": {
+					Value: 1,
+				},
+			},
+		},
+		{
+			Labels: agent.StringMap{
+				{Key: "service.name", Value: "apm-agent-ruby"},
+				{Key: "service.version", Value: "1.0"},
+			},
+			Samples: map[string]agent.Metric{
+				"processed_events": {
+					Value: 3,
+				},
+			},
+		},
+	}, metrics)
+}
+
+func TestServiceLimit(t *testing.T) {
+	counter := modelprocessor.NewServiceCounter()
+	batch := make(model.Batch, 20000)
+	for i := range batch {
+		service := model.Service{Name: genServiceName()}
+		batch[i] = model.APMEvent{Service: service}
+	}
+
+	err := counter.ProcessBatch(context.Background(), &batch)
+	require.NoError(t, err)
+
+	metrics := gatherMetrics(counter)[1:]
+	assert.Less(t, len(metrics), 10000)
```

Comment:
it seems like there's some sort of limit when gathering. I was receiving ~5000 metrics each time, even though I verified that internally the ServiceCounter was storing up to 10,000 service name/version pairs.

Reply:
Yeah, that was the issue, thanks!